### PR TITLE
[caffe2][nomnigraph] Batchnorm + Conv Fusion

### PR DESCRIPTION
**Pull request by bwasti**
moving this to an OSS diff

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -232,7 +232,7 @@ if (BUILD_PYTHON)
       caffe2_pybind11_state PROPERTIES LIBRARY_OUTPUT_DIRECTORY
       ${CMAKE_BINARY_DIR}/caffe2/python)
   target_link_libraries(
-      caffe2_pybind11_state caffe2_library)
+      caffe2_pybind11_state caffe2_library nomnigraph)
```
Comment (yinghai) [inline caffe2/CMakeLists.txt]: I guess this will be rebased away?

Comment (yinghai) [inline caffe2/CMakeLists.txt]: I guess this will be rebased away?

File: caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h [hunk 2]
```diff
@@ -475,6 +475,62 @@ class Receive : public NeuralNetOperator {
   string Source;
 };
 
+class BatchNormalization : public NeuralNetOperator {
```
Comment (yinghai) [inline caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h]: One general question is do we need to create a corresponding class on nomnigraph for c2 operator or we can just rely on the `OperatorDef*` attribute?

**Review by yinghai — APPROVED** [review]

File: caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h (hunk 2 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h]: It isn't strictly necessary to use nomnigraph, but the transformations need to know about that and use string comparison instead of isa<ClassName>() style checks (which are substantially more efficient).

I've tried to make it as seamless as possible with ops.def syntax for adding new classes, though

Comment (yinghai) [reply caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h]: Fair enough.

**Comment by yinghai** [comment]
There seem to be some legit compilation errors: 
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang3.8-ubuntu16.04-build/4041/console

**Comment by bwasti** [comment]
I'm reasonably confident these failures are related to the clang issue  (they exist within protobuf)
